v2/pkg/archive/strict_adder.go

```diff
@@ -0,0 +1,212 @@
+	}
+	err = o.tarWriter.Close()
+	if err != nil {
+		o.logger.Warn("error closing archive writer : %v", err)
+	}
+	return o.archiveFile.Close()
+}
+
+// addFile copies the contents of the `pathToFile` file from the disk into
+// the current chunk archive at `pathInTar` location.
+// addFile monitors the size of the current chunk, and creates a new chunk to
+// archive `pathToFile` if archiving would make the chunk exceed the maxArchiveSize.
+// addFile stops the archiving and returns an error if `pathToFile`'s size is
+// greater than maxArchiveSize.
+func (o *strictAdder) addFile(pathToFile string, pathInTar string) error {
+	fi, err := os.Stat(pathToFile)
+	if err != nil {
+		return err
+	}
+
+	// when a file is already bigger than the maxArchiveSize, it will not fit any chunk
+	// therefore we should stop
+	if fi.Size() > o.maxArchiveSize {
+		return fmt.Errorf("maxArchiveSize %dG is too small compared to sizes of files that need to be included in the archive.\n%s: %dG\n Aborting archive generation", o.maxArchiveSize/segMultiplier, fi.Name(), fi.Size()/segMultiplier)
+	}
+	// check if we should add this file to the archive without exceeding the maxArchiveSize
+	if fi.Size()+o.sizeOfCurrentChunk > o.maxArchiveSize {
+		err = o.nextChunk()
+		if err != nil {
+			return err
+		}
+	}
+	header, err := tar.FileInfoHeader(fi, fi.Name())
+	if err != nil {
+		return err
+	}
+	header.Name = pathInTar
+
+	if err := o.tarWriter.WriteHeader(header); err != nil {
+		return err
+	}
+	// Open the file for reading
+	file, err := os.Open(pathToFile)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	// Copy the file contents to the tar archive
+	if _, err := io.Copy(o.tarWriter, file); err != nil {
+		return err
+	}
+	o.sizeOfCurrentChunk += fi.Size()
+	return nil
+}
+
+// addAllFolder copies the contents of the `folderToAdd` from the disk into
+// the current chunk archive under `relativeTo` path.
+// addAllFolder monitors the size of the current chunk, and creates a new chunk to
+// archive files from `folderToAdd` if archiving would make the chunk exceed the maxArchiveSize.
+// addAllFolder stops the archiving and returns an error if the size of any file within the folder is
+// greater than maxArchiveSize.
+func (o *strictAdder) addAllFolder(folderToAdd string, relativeTo string) error {
+	return filepath.Walk(folderToAdd, func(path string, info os.FileInfo, incomingError error) error {
+		if incomingError != nil {
+			return incomingError
+		}
+		if info.IsDir() { // skip directories
+			return nil
+		}
+		// when a file is already bigger than the maxArchiveSize, it will not fit any chunk
+		// therefore we should stop
+		if info.Size() > o.maxArchiveSize {
+			return fmt.Errorf("maxArchiveSize %dG is too small compared to sizes of files that need to be included in the archive.\n%s: %dG\n Aborting archive generation", o.maxArchiveSize/segMultiplier, info.Name(), info.Size()/segMultiplier)
+		}
```

Comment:
same here with duplicate blocks lines 165-208 in permissive_adder.go , maybe create a common function for easier maintenance